### ISSUE TITLE
Loading a LoRA CrossEncoder checkpoint does not restore a trained `modules_to_save` classification head

### ISSUE THREAD

**Issue by shniubobo** [UKPLab/sentence-transformers#3980]
> [!NOTE]
>
> Generated by AI and edited by human

## Summary

Sentence Transformers can train a `CrossEncoder` with a PEFT LoRA adapter and then save the resulting adapter checkpoint. This checkpoint may include:

1. LoRA weights for selected transformer layers.
2. A separately trained sequence-classification head, configured through PEFT's `modules_to_save`.

For a sequence-classification model, the classification head is commonly named `score`. In this report, `score.weight` is the linear layer that converts the model's final hidden state into the relevance/classification score returned by the `CrossEncoder`.

When a checkpoint trained with `modules_to_save=["score"]` is loaded through:

```python
CrossEncoder("path/to/adapter-checkpoint")
```

the trained `score` weight is not restored into the module that PEFT actually uses during inference. Instead, that active classification head remains newly initialized. This means predictions use an untrained random head, even though the checkpoint contains the trained head weight.

The checkpoint loads with the following Transformers load report:

```text
score.weight                         | UNEXPECTED |
score.modules_to_save.default.weight | MISSING    |
```

Here:

- `UNEXPECTED` means a weight exists in the checkpoint but does not match a parameter expected by the model currently being loaded.
- `MISSING` means the model expects a parameter but did not receive a saved value for it, so Transformers initializes it from scratch.
- `score.modules_to_save.default.weight` is the active PEFT-managed copy of the classification head for the default adapter.

In this scenario, the `UNEXPECTED` entry is not safe to ignore, because it corresponds to the trained classification head and the `MISSING` parameter is the head actually used in forward passes.

### Environment

```text
sentence-transformers: 1d9647561b931252f868b9495172216161d26c82
peft: 0.20.0
transformers: 5.15.1
torch: 2.13.0+cu130
Python: 3.14
```

### How the checkpoint was created

The base model is a Hugging Face sequence-classification model used through Sentence Transformers' `CrossEncoder`.

The adapter was configured as follows:

```python
from peft import LoraConfig

lora_config = LoraConfig(
    r=8,
    lora_alpha=8,
    lora_dropout=0,
    target_modules="all-linear",
    modules_to_save=["score"],
    bias="none",
    task_type="SEQ_CLS",
)
```

`modules_to_save=["score"]` tells PEFT to train and save the complete `score` classification layer in addition to the LoRA weights.

The resulting `adapter_config.json` contains:

```json
{
  "modules_to_save": ["score"],
  "task_type": "SEQ_CLS"
}
```

The adapter checkpoint stores the trained classification-head tensor under this key:

```text
base_model.model.score.weight
```

This is the standard PEFT checkpoint key for a module listed in `modules_to_save`; PEFT maps it to the active adapter-specific copy when loading normally.

## Reproduction

```python
from pathlib import Path

import torch
from safetensors import safe_open
from sentence_transformers import CrossEncoder

checkpoint_path = Path("path/to/adapter-checkpoint")

# Read the trained classification head directly from the saved adapter.
with safe_open(
    checkpoint_path / "adapter_model.safetensors",
    framework="pt",
    device="cpu",
) as source:
    saved_score_weight = source.get_tensor(
        "base_model.model.score.weight"
    ).float()

# Load through Sentence Transformers' automatic checkpoint loading path.
encoder = CrossEncoder(
    checkpoint_path,
    model_kwargs={"torch_dtype": "float32"},
)

model = encoder.model
assert model is not None

# PEFT wraps modules listed in `modules_to_save`.
score = model.score

print(type(score))
print(score.active_adapters)
print(score.modules_to_save.keys())

# This should be True: the active inference head should equal the saved,
# trained classification head.
active_score_weight = (
    score.modules_to_save["default"].weight.detach().cpu().float()
)
print(torch.equal(active_score_weight, saved_score_weight))
```

## Actual behavior

During `CrossEncoder(checkpoint_path)` loading, Transformers reports:

```text
score.weight                         | UNEXPECTED |
score.modules_to_save.default.weight | MISSING    |
```

The model's `score` module is loaded as:

```text
peft.utils.other.ModulesToSaveWrapper
```

Its active adapter is `default`, and it contains:

```text
score.modules_to_save["default"].weight
```

However, this active weight does not equal the trained weight stored in the checkpoint:

```python
torch.equal(active_score_weight, saved_score_weight)
# False
```

The tensors differ substantially, rather than only by floating-point serialization noise. For example, the maximum absolute element-wise difference in one reproduction was approximately `0.095`; for normally restored `float32` checkpoint weights, this value should be exactly `0.0`.

### Why this changes inference results

PEFT's `ModulesToSaveWrapper` [uses the active adapter-specific copy in its forward path](https://github.com/huggingface/peft/blob/de5c7df7b1808546442b14e71a3e54a18370a679/src/peft/utils/other.py#L614-L617):

```python
def _forward_wrapped(self, *args, **kwargs):
    if not self.active_adapters:
        return self._forward_wrapped_passthrough(*args, **kwargs)
    return self.modules_to_save[self.active_adapters[0]](*args, **kwargs)
```

Therefore, with the active adapter set to `default`, inference uses:

```python
score.modules_to_save["default"].weight
```

rather than:

```python
score.original_module.weight
```

This was verified directly:

```python
inputs = torch.ones(2, score.original_module.in_features)

with torch.no_grad():
    baseline = score(inputs).clone()

    # Changing the original, inactive copy does not change forward output.
    score.original_module.weight.zero_()
    after_zeroing_original = score(inputs).clone()

    # Changing the active adapter copy does change forward output.
    score.modules_to_save["default"].weight.zero_()
    after_zeroing_active = score(inputs).clone()

print(torch.equal(baseline, after_zeroing_original))
# True

print(torch.equal(after_zeroing_original, after_zeroing_active))
# False
```

As a result, the automatically loaded model uses a newly initialized classification head instead of the head trained with the adapter.

## Expected behavior

For a checkpoint whose PEFT configuration contains:

```json
{
  "modules_to_save": ["score"]
}
```

`CrossEncoder(checkpoint_path)` should restore the checkpoint tensor:

```text
base_model.model.score.weight
```

into the active PEFT module:

```text
score.modules_to_save["default"].weight
```

After loading, this should hold:

```python
torch.equal(
    score.modules_to_save["default"].weight.detach().cpu().float(),
    saved_score_weight,
)
# True
```

The automatic Sentence Transformers loading path should produce the same model parameters and inference behavior as loading the base model and adapter through PEFT directly.

## Control experiment: the PEFT checkpoint is valid

The same checkpoint restores the trained classification head correctly when using PEFT directly:

```python
from pathlib import Path

import torch
from peft import PeftModel
from safetensors import safe_open
from transformers import AutoModelForSequenceClassification

base_model_path = Path("path/to/base-model")
adapter_path = Path("path/to/adapter-checkpoint")

with safe_open(
    adapter_path / "adapter_model.safetensors",
    framework="pt",
    device="cpu",
) as source:
    saved_score_weight = source.get_tensor(
        "base_model.model.score.weight"
    ).float()

base_model = AutoModelForSequenceClassification.from_pretrained(
    base_model_path,
    num_labels=1,
    torch_dtype=torch.float32,
)

model = PeftModel.from_pretrained(base_model, adapter_path)

active_score_weight = (
    model.base_model.model.score.modules_to_save["default"]
    .weight.detach()
    .cpu()
    .float()
)

print(torch.equal(active_score_weight, saved_score_weight))
# True

print((active_score_weight - saved_score_weight).abs().max().item())
# 0.0
```

This indicates that the adapter checkpoint format is valid and that PEFT can restore it correctly. The failure appears specific to Sentence Transformers' automatic adapter-loading path.

## Suggested fix

The automatic loading path should avoid treating `adapter_model.safetensors` as an ordinary full-model checkpoint before the PEFT adapter structure has been created.

A possible approach is:

1. Load the base sequence-classification model from the adapter configuration's `base_model_name_or_path`.
2. Detect that the checkpoint is a PEFT adapter checkpoint, for example by the presence of `adapter_config.json`.
3. Load the adapter using PEFT's native API:

   ```python
   PeftModel.from_pretrained(base_model, adapter_checkpoint_path)
   ```

4. Construct or return the `CrossEncoder` around that PEFT-wrapped model.

Alternatively, if Sentence Transformers must retain its existing loading flow, it should translate the saved PEFT `modules_to_save` key:

```text
base_model.model.score.weight
```

to the key expected after wrapper creation:

```text
score.modules_to_save.default.weight
```

and ensure the PEFT wrapper is installed before applying the adapter state dict.

A regression test should train or construct a small sequence-classification LoRA adapter with `modules_to_save=["score"]`, save it, reload it through `CrossEncoder(checkpoint_path)`, and verify:

```python
loaded_score.modules_to_save["default"].weight == saved_adapter_score_weight
```

and identical inference logits between the Sentence Transformers automatic-loading path and the equivalent PEFT-native loading path.

**Comment by tomaarsen** [comment]
Hello!

Yes, this was reported in #3978. In short, with `modules_to_save=["classifier"]` you can save classifier weights, but if your base model doesn't have any classifier weights to begin with, it'll randomize those anew before the saved PEFT weights are added on top. This is likely something that should be fixed in `transformers` and/or `peft`, i.e. that 1) loading a `AutoModelFor...` instance (i.e. a body + a head) from a checkpoint that only has a body (e.g. `https://huggingface.co/google-bert/bert-base-uncased`) or a different head (e.g. loading a causal-lm model as sequence classification), then 2) adding a PEFT Adapter, then 3) saving the model, and then 4) reloading the model fails as the head weights are randomized before the PEFT Adapter weights are added. This issue goes beyond Sentence Transformers, we just notice it with the `CrossEncoder`.

Granted: Perhaps PEFT is simply not a good option for loading a model with a different head than the original model was trained for, as the new head weights might require "full" weight updates as the weights are untrained upon loading.

- Tom Aarsen

**Comment by shniubobo** [comment]
>  This is likely something that should be fixed in `transformers` and/or `peft`

I'm not quite sure if this is true. If you try running the code in the "Control experiment" part, which uses only the upstream `transformers` and `peft`, the same issue does not occur.

> i.e. that 1) loading a AutoModelFor... instance (i.e. a body + a head) from a checkpoint that only has a body (e.g. `https://huggingface.co/google-bert/bert-base-uncased`) or a different head (e.g. loading a causal-lm model as sequence classification), then 2) adding a PEFT Adapter, then 3) saving the model, and then 4) reloading the model fails as the head weights are randomized before the PEFT Adapter weights are added.

The control experiment basically succeeds in achieving the 4th step by repeating the first three steps. Reloading the model does give a warning:

```
Key          | Status  | 
-------------+---------+-
score.weight | MISSING | 

Notes:
- MISSING:      those params were newly initialized because missing from the checkpoint. Consider training on your downstream task.
```

But that is expected, and after we load the adapter, we do get back the trained classification weights we want.

> Granted: Perhaps PEFT is simply not a good option for loading a model with a different head than the original model was trained for, as the new head weights might require "full" weight updates as the weights are untrained upon loading.

Perhaps I'm using PEFT in a wrong way, but I think PEFT does support full weight updates of layers specified in `modules_to_save`.

With the `LoraConfig` in my original post, and with this:

```python
[
    (name, parameter.numel())
    for name, parameter in cross_encoder.model.named_parameters()
    if parameter.requires_grad
]
```

..., where `cross_encoder` is a `CrossEncoder` instance, I get:

```python
[
    ('model.layers.0.linear_attn.out_proj.lora_A.default.weight', 16384),
    ('model.layers.0.linear_attn.out_proj.lora_B.default.weight', 8192),
    # ...
    ('score.modules_to_save.default.weight', 1024),
]
```

The hidden dimension of the base model is 1024, and the classification head has one label, so `('score.modules_to_save.default.weight', 1024)` means the head is being fully fine-tuned.

**Comment by tomaarsen** [comment]
Sentence Transformers doesn't load adapters with
```python
base_model = AutoModelForSequenceClassification.from_pretrained(
    base_model_path,
    num_labels=1,
    torch_dtype=torch.float32,
)

model = PeftModel.from_pretrained(base_model, adapter_path)
```
It loads with:
```python
model = AutoModelForSequenceClassification.from_pretrained(
    adapter_path
)
```

---

If I try the setup with loading a normal model, adding an adapter, saving the model, and then loading in the same way that ST does, then it breaks because `transformers` or `peft` randomizes the new weights for the head before adding the adapter weights.
```python
import torch
from transformers import AutoModelForSequenceClassification
from peft import LoraConfig

# 1. Load a model without a freshly initialized head + apply a LoRA config
model = AutoModelForSequenceClassification.from_pretrained(
    "google-bert/bert-base-uncased",
    num_labels=1,
    torch_dtype=torch.float32,
)
lora_config = LoraConfig(
    r=8,
    lora_alpha=8,
    lora_dropout=0,
    target_modules="all-linear",
    modules_to_save=["classifier"],
    bias="none",
    task_type="SEQ_CLS",
)
model.add_adapter(lora_config)
# Grab the weights of the classifier before saving the model
applied_classifier_weight = model.classifier.weight.data

# Save so we can load this adapter directly
model.save_pretrained("bert-base-adapter")

# 2. Load the model with the adapter applied
adapter_model = AutoModelForSequenceClassification.from_pretrained("bert-base-adapter")
loaded_classifier_weight = adapter_model.classifier.weight.data

# Compare the weights of the classifier before and after saving/loading
print(torch.equal(applied_classifier_weight, loaded_classifier_weight))
# False
print((applied_classifier_weight - loaded_classifier_weight).abs().max().item())
# 0.10951237380504608
```

If I instead do the exact same for a model that uses the same head (e.g. sequence classification), then there's no issue instead:

```python
import torch
from transformers import AutoModelForSequenceClassification
from peft import LoraConfig

# 1. Load a model without a freshly initialized head + apply a LoRA config
model = AutoModelForSequenceClassification.from_pretrained(
    "BAAI/bge-reranker-base",
    num_labels=1,
    torch_dtype=torch.float32,
)
lora_config = LoraConfig(
    r=8,
    lora_alpha=8,
    lora_dropout=0,
    target_modules="all-linear",
    modules_to_save=["classifier"],
    bias="none",
    task_type="SEQ_CLS",
)
model.add_adapter(lora_config)
# Grab the weights of the classifier before saving the model
applied_classifier_weight = model.classifier.dense.weight.data

# Save so we can load this adapter directly
model.save_pretrained("bge-reranker-base-adapter")

# 2. Load the model with the adapter applied
adapter_model = AutoModelForSequenceClassification.from_pretrained("bert-base-adapter")
loaded_classifier_weight = adapter_model.classifier.dense.weight.data

# Compare the weights of the classifier before and after saving/loading
print(torch.equal(applied_classifier_weight, loaded_classifier_weight))
# True
print((applied_classifier_weight - loaded_classifier_weight).abs().max().item())
# 0.0
```

- Tom Aarsen

**Comment by shniubobo** [comment]
> If I instead do the exact same for a model that uses the same head (e.g. sequence classification), then there's no issue instead:

Then one workaround for my use case (initializing a `AutoModelForSequenceClassification` from a `...ForConditionalGeneration` model) is to randomly initialize a classification head (with `AutoModelForSequenceClassification`), save the initialized model seperately, and then load from it instead. But this is at best a workaround, as it doesn't seem to me quite ergonomic having to make a copy of the entire model just to add a randomly initialized layer.

---

> Sentence Transformers doesn't load adapters with
> 
> ```python
> base_model = AutoModelForSequenceClassification.from_pretrained(
>     base_model_path,
>     num_labels=1,
>     torch_dtype=torch.float32,
> )
> 
> model = PeftModel.from_pretrained(base_model, adapter_path)
> ```
>
> It loads with:
> 
> ```python
> model = AutoModelForSequenceClassification.from_pretrained(
>     adapter_path
> )
> ```

Another workaround could be taking the first approach in ST, but that would require specifying kwargs for `AutoModelForSequenceClassification.from_pretrained`, which might not be feasible for the current code flow.

---

Ultimately, an upstream fix would be more preferable, by making the same-head and different-head situations behave the same. I will look into `transformers` and `peft` to see where the fix fits best, and report back once it gets merged. In the meanwhile, please leave this issue open.

**Comment by tomaarsen** [comment]
> Another workaround could be taking the first approach in ST, but that would require specifying kwargs for `AutoModelForSequenceClassification.from_pretrained`, which might not be feasible for the current code flow.

Previously, both were semi-supported, but as of recently, I've moved to only supporting the bottom option via the `transformers` PEFT integration. Ideally I'd like it to stay that way. I agree that a transformers / peft fix might be nicest.

- Tom Aarsen

**Comment by shniubobo** [comment]
I've created a PR to `transformers`, which should hopefully fix the issue. Please see the linked PR.

---

> If I instead do the exact same for a model that uses the same head (e.g. sequence classification), then there's no issue instead

FYI this was a false negative. To cite my PR:

> **If the base model has a classification head**, the head loaded from the LoRA adapter is discarded, and the base model's pretrained head is used

To verify, run:

```python
import torch
from peft import LoraConfig
from transformers import AutoModelForSequenceClassification

# 1. Load a model without a freshly initialized head + apply a LoRA config
model = AutoModelForSequenceClassification.from_pretrained(
    "BAAI/bge-reranker-base",
    num_labels=1,
    torch_dtype=torch.float32,
)
lora_config = LoraConfig(
    r=8,
    lora_alpha=8,
    lora_dropout=0,
    target_modules="all-linear",
    modules_to_save=["classifier"],
    bias="none",
    task_type="SEQ_CLS",
)
model.add_adapter(lora_config)
# Mutate the weights as though we were training it
with torch.no_grad():
    model.classifier.modules_to_save.default.dense.weight.fill_(0.1234)
# Grab the weights of the classifier before saving the model
applied_classifier_weight = model.classifier.modules_to_save.default.dense.weight

# Save so we can load this adapter directly
model.save_pretrained("bge-reranker-base-adapter")

# 2. Load the model with the adapter applied
adapter_model = AutoModelForSequenceClassification.from_pretrained(
    "bge-reranker-base-adapter"
)
loaded_classifier_weight = adapter_model.classifier.modules_to_save.default.dense.weight

# Compare the weights of the classifier before and after saving/loading
print(torch.equal(applied_classifier_weight, loaded_classifier_weight))
# False
print((applied_classifier_weight - loaded_classifier_weight).abs().max().item())
# 0.22564032673835754
```

So there is actually no difference between the same-head and different-head situations.
